blog-post/blog-post.md

```diff
@@ -0,0 +1,634 @@
+machinery for ABI encoding of values.
+
+```
+forall ty . ty : ABIAttribs, ty : ABIEncode => function log(val : ty) {
+    let CONSOLE_ADDRESS : word = 0x000000000000000000636F6e736F6c652e6c6f67;
+    let payload = abi_encode(val);
+
+    // extract the underlying word representation of the payload
+    let ptr = Typedef.rep(payload);
+
+    assembly {
+        pop(
+            staticcall(
+                gas(),
+                CONSOLE_ADDRESS,
+                add(ptr, 32),
+                mload(ptr),
+                0,
+                0
+            )
+        )
+    }
+}
+```
+
+Another short possible implementation which relies on EVM `log1` instruction,
+instead of Foundry console address, would be:
+
+```
+forall t . t : Typedef(word) => function log1(v : t, topic : word) -> () {
+    let w : word = Typedef.rep(v);
+    assembly {
+        mstore(0,w)
+        log1(0,32,topic)
+    }
+}
+```
+
+This version uses standard library class `Typedef`, which provides functions for
+the conversion between `word` and types which are instance of this class.
+
+### Algebraic data types and pattern matching
+
+Algebraic Data Types provide a way of data modeling using sum types
+(disjoint unions) and product types (structural records). This enables the construction of
+precise types where invalid states are **unrepresentable** by design, thereby enhancing program
+correctness through the type system itself. As an example, consider the following type
+definition for an auction state.
+
+```
+data AuctionState =
+    NotStarted (word)
+  | Active (word, address)
+  | Ended (word, address)
+  | Cancelled;
+
+```
+The type has three constructors: `NotStarted` specifies that the auction has not started yet and
+stores its reserved price, `Active` denotes that the auction has begun and it stores the
+current highest bid and the address that made such bid, `Ended` represents that the auction has
+finished with success and it holds the highest bid and the winner address and constructor,
+`Cancelled` is used when the auction has been cancelled.
+
+Using algebraic data types, we can define functions by pattern matching. As an example,
+consider function `processAuction` which tries to update the action state based on current
+state and `msg.value`.  Pattern matching provides structural decomposition of data types through
+case analysis. This ensures all possible variants are handled explicitly, eliminating
+partial function hazards and providing formal guarantees of match completeness.
+
+```
+function processAuction(state) {
+    match state {
+    | NotStarted(reserve) =>
+        require(msg.value >= reserve);
+        return Active(msg.value, msg.sender);
+    | Active(currentBid, bidder) =>
+        require(msg.value > currentBid);
+        transferFunds(bidder, currentBid);
+        return Active(msg.value, msg.sender);
+    | Cancelled =>
+        revert();
+    |   _ =>
+        return state;
```

Comment:
Actually, doesn't `_` defeat the exhaustiveness check? The text above says "This ensures all possible variants are handled explicitly", but with this you can define a new variant and not handle it at all.